File: UdidServer/util/src/com/horizon/util/MHash.java
package com.horizon.util;


/**
 * murmurhash
 *
 * see:
 * https://sites.google.com/site/murmurhash/
 *
 */
public class MHash {
    public static long hash64(String str) {
        if (str == null || str.length() == 0) {
            return 0L;
        }
        byte[] bytes = str.getBytes();
        return hash64(bytes, bytes.length);
    }

    public static long hash64(final byte[] data, int len) {
        if (data == null || data.length == 0 || len == 0) {
            return 0L;
        }
        final long m = 0xc6a4a7935bd1e995L;
        final long seed = 0xe17a1465L;
        final int r = 47;

        long h = seed ^ (len * m);
        int remain = len & 7;
        int size = len - remain;

        for (int i = 0; i < size; i += 8) {
            long k = ((long) data[i+7] << 56) +
                    ((long) (data[i + 6] & 0xFF) << 48) +
                    ((long) (data[i + 5] & 0xFF) << 40) +
                    ((long) (data[i + 4] & 0xFF) << 32) +
                    ((long) (data[i + 3] & 0xFF) << 24) +
                    ((data[i + 2] & 0xFF) << 16) +
                    ((data[i + 1] & 0xFF) << 8) +
                    ((data[i] & 0xFF));
            k *= m;
            k ^= k >>> r;
            k *= m;
            h ^= k;
            h *= m;
        }

        switch (remain) {
            case 7:
                h ^= (long) (data[size + 6] & 0xFF) << 48;
            case 6:
                h ^= (long) (data[size + 5] & 0xFF) << 40;
            case 5:
                h ^= (long) (data[size + 4] & 0xFF) << 32;
            case 4:
                h ^= (long) (data[size + 3] & 0xFF) << 24;
            case 3:
                h ^= (data[size + 2] & 0xFF) << 16;
            case 2:
                h ^= (data[size + 1] & 0xFF) << 8;
            case 1:
                h ^= (data[size] & 0xFF);
                h *= m;
        }

        h ^= h >>> r;
        h *= m;
        h ^= h >>> r;

        return h;
    }
}

File: UdidServer/src/main/kotlin/horizon/util/MHash.kt
package horizon.util

/**
 * murmurhash
 *
 * see:
 * https://sites.google.com/site/murmurhash/
 *
 */
object MHash {
    fun hash64(str: String?): Long {
        if (str == null || str.length == 0) {
            return 0L
        }
        val bytes = str.toByteArray()
        return hash64(bytes, bytes.size)
    }

    fun hash64(data: ByteArray?, len: Int): Long {
        if (data == null || data.size == 0 || len == 0) {
            return 0L
        }
        val m = -0x395b586ca42e166bL
        val seed = 0xe17a1465L
        val r = 47
        var h = seed xor len * m
        val remain = len and 7
        val size = len - remain
        var i = 0
        while (i < size) {
            var k = (data[i + 7].toLong() shl 56) +
                    ((data[i + 6].toInt() and 0xFF).toLong() shl 48) +
                    ((data[i + 5].toInt() and 0xFF).toLong() shl 40) +
                    ((data[i + 4].toInt() and 0xFF).toLong() shl 32) +
                    ((data[i + 3].toInt() and 0xFF).toLong() shl 24) +
                    (data[i + 2].toInt() and 0xFF shl 16) +
                    (data[i + 1].toInt() and 0xFF shl 8) +
                    (data[i].toInt() and 0xFF)
            k *= m
            k = k xor (k ushr r)
            k *= m
            h = h xor k
            h *= m
            i += 8
        }
        when (remain) {
            7 -> {
                h = h xor ((data[size + 6].toInt() and 0xFF).toLong() shl 48)
                h = h xor ((data[size + 5].toInt() and 0xFF).toLong() shl 40)
                h = h xor ((data[size + 4].toInt() and 0xFF).toLong() shl 32)
                h = h xor ((data[size + 3].toInt() and 0xFF).toLong() shl 24)
                h = h xor (data[size + 2].toInt() and 0xFF shl 16).toLong()
                h = h xor (data[size + 1].toInt() and 0xFF shl 8).toLong()
                h = h xor (data[size].toInt() and 0xFF).toLong()
                h *= m
            }

            6 -> {
                h = h xor ((data[size + 5].toInt() and 0xFF).toLong() shl 40)
                h = h xor ((data[size + 4].toInt() and 0xFF).toLong() shl 32)
                h = h xor ((data[size + 3].toInt() and 0xFF).toLong() shl 24)
                h = h xor (data[size + 2].toInt() and 0xFF shl 16).toLong()
                h = h xor (data[size + 1].toInt() and 0xFF shl 8).toLong()
                h = h xor (data[size].toInt() and 0xFF).toLong()
                h *= m
            }

            5 -> {
                h = h xor ((data[size + 4].toInt() and 0xFF).toLong() shl 32)
                h = h xor ((data[size + 3].toInt() and 0xFF).toLong() shl 24)
                h = h xor (data[size + 2].toInt() and 0xFF shl 16).toLong()
                h = h xor (data[size + 1].toInt() and 0xFF shl 8).toLong()
                h = h xor (data[size].toInt() and 0xFF).toLong()
                h *= m
            }

            4 -> {
                h = h xor ((data[size + 3].toInt() and 0xFF).toLong() shl 24)
                h = h xor (data[size + 2].toInt() and 0xFF shl 16).toLong()
                h = h xor (data[size + 1].toInt() and 0xFF shl 8).toLong()
                h = h xor (data[size].toInt() and 0xFF).toLong()
                h *= m
            }

            3 -> {
                h = h xor (data[size + 2].toInt() and 0xFF shl 16).toLong()
                h = h xor (data[size + 1].toInt() and 0xFF shl 8).toLong()
                h = h xor (data[size].toInt() and 0xFF).toLong()
                h *= m
            }

            2 -> {
                h = h xor (data[size + 1].toInt() and 0xFF shl 8).toLong()
                h = h xor (data[size].toInt() and 0xFF).toLong()
                h *= m
            }

            1 -> {
                h = h xor (data[size].toInt() and 0xFF).toLong()
                h *= m
            }
        }
        h = h xor (h ushr r)
        h *= m
        h = h xor (h ushr r)
        return h
    }
}
